core/src/avm2/globals/flash/utils/bytearray.rs

```diff
@@ -0,0 +1,946 @@
+    }
+
+    Ok(Value::Undefined)
+}
+
+/// Implements `flash.utils.ByteArray`'s class constructor.
+pub fn class_init<'gc>(
+    _activation: &mut Activation<'_, 'gc, '_>,
+    _this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    Ok(Value::Undefined)
+}
+
+/// Writes a single byte to the bytearray
+pub fn write_byte<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        let byte = args
+            .get(0)
+            .cloned()
+            .unwrap_or(Value::Undefined)
+            .coerce_to_i32(activation)?;
+        bytearray.write_byte(byte as u8);
+    }
+
+    Ok(Value::Undefined)
+}
+
+/// Writes multiple bytes to the bytearray from another bytearray
+pub fn write_bytes<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(Value::Object(second_array)) = args.get(0) {
+        let combining_bytes = second_array.as_bytearray().unwrap().bytes().clone();
+
+        let offset = args
+            .get(1)
+            .unwrap_or(&Value::Unsigned(0))
+            .coerce_to_u32(activation)? as usize;
+        let length = args
+            .get(2)
+            .unwrap_or(&Value::Unsigned(0))
+            .coerce_to_u32(activation)? as usize;
+
+        // In the docs it says "If offset or length is out of range, they are clamped to the beginning and end of the bytes array."
+        // However, in the actual flash player, it seems to just raise an error.
+        if offset + length > combining_bytes.len() {
+            log::error!("ByteArray: Reached EOF");
+            return Ok(Value::Undefined);
+        }
+
+        if let Some(mut bytearray) = this
+            .unwrap()
+            .as_bytearray_mut(activation.context.gc_context)
+        {
+            bytearray.write_bytes(if length != 0 {
+                &combining_bytes[offset..length + offset]
+            } else {
+                &combining_bytes[offset..]
+            });
+        }
+    }
+
+    Ok(Value::Undefined)
+}
+
+// Reads the bytes from the current bytearray into another bytearray
+pub fn read_bytes<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    let current_bytes = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+        .unwrap()
+        .bytes()
+        .clone();
+    let position = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+        .unwrap()
+        .position();
+    let mut merging_offset = 0;
+    if let Some(Value::Object(second_array)) = args.get(0) {
+        let offset = args
+            .get(1)
```

Comment:
A large amount of these are unwraps that could fail if some other invariant elsewhere in the interpreter, say, called this function without a `this` parameter, or with a non-ByteArray `this`, or with the wrong arguments. None of that is well-enforced in our interpreter yet. Most of these unwraps should be replaced with equivalents (e.g. returning an `ArgumentError`-like string)